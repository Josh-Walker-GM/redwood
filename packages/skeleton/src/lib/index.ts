import fs from 'fs'

import { parse } from '@babel/parser'

export function getFileAST(filePath: string) {
  const code = fs.readFileSync(filePath, { encoding: 'utf8', flag: 'r' })
  return parse(code, {
    sourceType: 'unambiguous',
    plugins: ['jsx', 'typescript'],
  })
}
