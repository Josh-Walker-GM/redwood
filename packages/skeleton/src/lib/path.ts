import path from 'path'

import { getBaseDir, getConfigPath } from '@redwoodjs/internal/dist/index'

/**
 * Determines the root path of a redwood project which contains the provided path
 * @param pathWithinProject A path within a redwood project
 * @returns The root path of the associated project
 */
export function getRootPath(pathWithinProject: string) {
  return pathWithinProject
    ? path.dirname(getConfigPath(pathWithinProject))
    : getBaseDir()
}
