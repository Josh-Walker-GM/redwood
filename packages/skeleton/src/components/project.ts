import fs from 'fs'
import path from 'path'

import chalk from 'chalk'

import { getPaths } from '@redwoodjs/internal/dist/index'

import { getRootPath } from '../lib/path'

import { RedwoodSkeleton } from './base'
import { extractCells, RedwoodCell } from './cell'
import { extractFunctions, RedwoodFunction } from './function'
import { extractLayouts, RedwoodLayout } from './layout'
import { extractPages, RedwoodPage } from './page'
import { extractRouters, RedwoodRouter } from './router'
import { extractSides, RedwoodSide } from './side'

/**
 * Used to enumerate either JS or TS project types
 */
export enum RedwoodProjectType {
  TYPESCRIPT = 'typescript',
  JAVASCRIPT = 'javascript',
}

export class RedwoodProject extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  readonly type: RedwoodProjectType

  private sides?: RedwoodSide[] | undefined
  private cells?: RedwoodCell[] | undefined
  private routers?: RedwoodRouter[] | undefined
  private layouts?: RedwoodLayout[] | undefined
  private pages?: RedwoodPage[] | undefined
  private functions?: RedwoodFunction[] | undefined

  public static getProject({
    pathWithinProject = '',
    full = false,
    readFromCache = true,
    insertIntoCache = true,
  }: {
    pathWithinProject?: string
    full?: boolean
    readFromCache?: boolean
    insertIntoCache?: boolean
  } = {}) {
    const projectCache = RedwoodProjectsCache.getInstance().projects
    const rootPath = getRootPath(pathWithinProject)

    if (readFromCache) {
      const cachedProject = projectCache.get(rootPath)
      if (cachedProject) {
        return cachedProject
      }
    }

    const project = new RedwoodProject(rootPath, full)
    if (insertIntoCache) {
      projectCache.set(project.filepath, project)
    }
    return project
  }

  private constructor(rootPath: string, full = false) {
    super(rootPath)

    // A project is typescript if we detect a tsconfig.json
    this.type =
      fs.existsSync(
        path.join(getPaths(this.filepath).web.base, 'tsconfig.json')
      ) ||
      fs.existsSync(
        path.join(getPaths(this.filepath).api.base, 'tsconfig.json')
      )
        ? RedwoodProjectType.TYPESCRIPT
        : RedwoodProjectType.JAVASCRIPT

    if (full) {
      this.sides = extractSides(this)
      this.cells = extractCells(this)
      this.layouts = extractLayouts(this)
      this.pages = extractPages(this)
      this.functions = extractFunctions(this)
      this.routers = extractRouters(this)
    }
  }

  getSides(forceExtract = false): RedwoodSide[] {
    if (forceExtract) {
      return extractSides(this)
    }
    return this.sides === undefined ? extractSides(this) : this.sides
  }

  getCells(forceExtract = false): RedwoodCell[] {
    if (forceExtract) {
      return extractCells(this)
    }
    return this.cells === undefined ? extractCells(this) : this.cells
  }

  getRouters(forceExtract = false): RedwoodRouter[] {
    if (forceExtract) {
      return extractRouters(this)
    }
    return this.routers === undefined ? extractRouters(this) : this.routers
  }

  getLayouts(forceExtract = false): RedwoodLayout[] {
    if (forceExtract) {
      return extractLayouts(this)
    }
    return this.layouts === undefined ? extractLayouts(this) : this.layouts
  }

  getPages(forceExtract = false): RedwoodPage[] {
    if (forceExtract) {
      return extractPages(this)
    }
    return this.pages === undefined ? extractPages(this) : this.pages
  }

  getFunctions(forceExtract = false): RedwoodFunction[] {
    if (forceExtract) {
      return extractFunctions(this)
    }
    return this.functions === undefined
      ? extractFunctions(this)
      : this.functions
  }

  // Diagnostics

  getInformation(): string {
    return '' // TODO: Implement, gets the info of all other components too
  }

  printInformation(): void {
    console.log(
      `${chalk.bgCyan('[Info]')}\t${this.name}\t${chalk.dim(this.filepath)}`
    )
  }

  printWarnings(cascade = false): void {
    if (this.warnings.length > 0) {
      const titleLine = `${chalk.bgYellow('[Warn]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const warningLines = this.warnings.map((warning, index) => {
        return ` (${index + 1}) ${warning}\n`
      })
      console.log(titleLine.concat('\n', ...warningLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printWarnings()
      })
      this.functions?.forEach((func) => {
        func.printWarnings()
      })
      this.layouts?.forEach((layout) => {
        layout.printWarnings()
      })
      this.pages?.forEach((page) => {
        page.printWarnings()
      })
      this.routers?.forEach((router) => {
        router.printWarnings()
        router.routes.forEach((route) => {
          route.printWarnings()
        })
      })
      this.sides?.forEach((side) => {
        side.printWarnings()
      })
    }
  }

  printErrors(cascade = false): void {
    if (this.errors.length > 0) {
      const titleLine = `${chalk.bgRed('[Error]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const errorLines = this.errors.map((error, index) => {
        return ` (${index + 1}) ${error}\n`
      })
      console.log(titleLine.concat('\n', ...errorLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printErrors()
      })
      this.functions?.forEach((func) => {
        func.printErrors()
      })
      this.layouts?.forEach((layout) => {
        layout.printErrors()
      })
      this.pages?.forEach((page) => {
        page.printErrors()
      })
      this.routers?.forEach((router) => {
        router.printErrors()
        router.routes.forEach((route) => {
          route.printErrors()
        })
      })
      this.sides?.forEach((side) => {
        side.printErrors()
      })
    }
  }
}

/**
 * Singelton class which contains a cache of {@link RedwoodProject} instances.
 *
 * This class is not exported and should only be available from static methods within {@link RedwoodProject}, in order to keep the surface of the skeleton API cleaner and leaner.
 */
class RedwoodProjectsCache {
  private static instance: RedwoodProjectsCache

  public projects: Map<string, RedwoodProject>

  private constructor() {
    this.projects = new Map()
  }

  public static getInstance(): RedwoodProjectsCache {
    if (!RedwoodProjectsCache.instance) {
      RedwoodProjectsCache.instance = new RedwoodProjectsCache()
    }
    return RedwoodProjectsCache.instance
  }
}
