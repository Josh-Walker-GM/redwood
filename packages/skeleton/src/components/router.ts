import fs from 'fs'
import path from 'path'

import { RedwoodSkeleton } from './skeleton'
import { RedwoodProject } from './project'
import { extractRoutes } from './route'
import type { RedwoodRoute } from './route'
import { extractSides, RedwoodSideType } from './side'
import type { RedwoodSide } from './side'

export class RedwoodRouter extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  readonly routes: RedwoodRoute[]

  constructor(filepath: string) {
    super(filepath)

    this.routes = extractRoutes(this)

    // Checks

    // Not found route checks
    const notFoundRoutes = this.routes.filter((route) => {
      return route.isNotFound
    })
    if (notFoundRoutes.length === 0) {
      this.errors.push('No notfound route detected')
    } else if (notFoundRoutes.length > 1) {
      this.errors.push('No more than one notfound route should be present')
    }

    // Duplicate routes checking
    const nameOccurences: Record<string, number> = {}
    this.routes.forEach((route) => {
      if (route.name in nameOccurences && nameOccurences[route.name] < 2) {
        this.errors.push(`Multiple routes named ${route.name} are present`)
        nameOccurences[route.name] += 1
      } else {
        nameOccurences[route.name] = 1
      }
    })
  }

  getSide(): RedwoodSide {
    const sides = RedwoodProject.getProject({
      pathWithinProject: this.filepath,
    }).getSides()
    const side = sides.find((side) => {
      return this.filepath.startsWith(side.filepath)
    })
    if (side) {
      return side
    }
    throw new Error('Could not determine which side the router belongs to.')
  }
}

export function extractRouter(filepath: string): RedwoodRouter {
  return new RedwoodRouter(filepath)
}

export function extractRouters(
  project: RedwoodProject | undefined = undefined
): RedwoodRouter[] {
  const routers: RedwoodRouter[] = []

  const routerFiles: string[] = []
  const sides = project ? project.getSides(true) : extractSides(undefined)
  sides
    ?.filter((side) => {
      // Extract only sides which support a router
      return side.type === RedwoodSideType.WEB
    })
    .forEach((side) => {
      // Find the router file and create a RedwoodRouter
      let routerFileName = ''
      switch (side.type) {
        case RedwoodSideType.WEB:
          routerFileName = fs
            .readdirSync(path.join(side.filepath, 'src'))
            .filter((path) => {
              return path.match(/Routes\.(js|jsx|tsx)$/)
            })[0]
          // TODO: fix: Assumes the router file exists
          routerFiles.push(path.join(side.filepath, 'src', routerFileName))
          break
        default:
          // TODO: Determine how to handle this
          break
      }
    })

  routerFiles.forEach((routerFile) => {
    routers.push(new RedwoodRouter(routerFile))
  })

  return routers
}
