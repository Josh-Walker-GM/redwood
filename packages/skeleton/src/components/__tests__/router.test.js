import path from 'path'

import { stripAndFormatPathForTesting } from '../../lib/testing'
import { RedwoodProject } from '../project'
import { extractRouters } from '../router'

const FIXTURE_PATH = path.join(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '..',
  '__fixtures__'
)

describe.each([
  'empty-project',
  'example-todo-main',
  'example-todo-main-with-errors',
  'test-project',
])('From within the %s fixture', (PROJECT_NAME) => {
  const PROJECT_PATH = path.join(FIXTURE_PATH, PROJECT_NAME)

  const RWJS_CWD = process.env.RWJS_CWD
  beforeAll(() => {
    process.env.RWJS_CWD = PROJECT_PATH
  })
  afterAll(() => {
    process.env.RWJS_CWD = RWJS_CWD
  })

  it('returns the correct routers without a project', () => {
    const routers = extractRouters(undefined)
    routers.forEach((router) => {
      router.filepath = stripAndFormatPathForTesting(
        router.filepath,
        PROJECT_PATH
      )
      router.routes.forEach((route) => {
        route.filepath = stripAndFormatPathForTesting(
          route.filepath,
          PROJECT_PATH
        )
      })
    })
    expect(routers).toMatchSnapshot()
  })

  it('returns the correct routers with a project', () => {
    const project = RedwoodProject.getProject({
      pathWithinProject: PROJECT_PATH,
      readFromCache: false,
      insertIntoCache: false,
    })
    const routers = extractRouters(project)
    routers.forEach((router) => {
      router.filepath = stripAndFormatPathForTesting(
        router.filepath,
        PROJECT_PATH
      )
      router.routes.forEach((route) => {
        route.filepath = stripAndFormatPathForTesting(
          route.filepath,
          PROJECT_PATH
        )
      })
    })
    expect(routers).toMatchSnapshot()
  })
})
