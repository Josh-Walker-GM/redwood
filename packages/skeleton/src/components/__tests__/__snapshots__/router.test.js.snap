// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the empty-project fixture returns the correct routers with a project 1`] = `
[
  RedwoodRouter {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "name": "Routes",
    "routes": [
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": true,
        "isPrivate": false,
        "name": "notfound",
        "pageIdentifier": "NotFoundPage",
        "path": undefined,
        "prerender": false,
        "warnings": [],
      },
    ],
    "warnings": [],
  },
]
`;

exports[`From within the empty-project fixture returns the correct routers without a project 1`] = `
[
  RedwoodRouter {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "name": "Routes",
    "routes": [
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": true,
        "isPrivate": false,
        "name": "notfound",
        "pageIdentifier": "NotFoundPage",
        "path": undefined,
        "prerender": false,
        "warnings": [],
      },
    ],
    "warnings": [],
  },
]
`;

exports[`From within the example-todo-main fixture returns the correct routers with a project 1`] = `
[
  RedwoodRouter {
    "errors": [],
    "filepath": "/web/src/Routes.js",
    "name": "Routes",
    "routes": [
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "home",
        "pageIdentifier": "HomePage",
        "path": "/",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "typescriptPage",
        "pageIdentifier": "TypeScriptPage",
        "path": "/typescript",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "someOtherPage",
        "pageIdentifier": "EditUserPage",
        "path": "/somewhereElse",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "fooPage",
        "pageIdentifier": "FooPage",
        "path": "/foo",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "barPage",
        "pageIdentifier": "BarPage",
        "path": "/bar",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": true,
        "name": "privatePage",
        "pageIdentifier": "PrivatePage",
        "path": "/private-page",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": true,
        "isPrivate": false,
        "name": "notfound",
        "pageIdentifier": "NotFoundPage",
        "path": undefined,
        "prerender": false,
        "warnings": [],
      },
    ],
    "warnings": [],
  },
]
`;

exports[`From within the example-todo-main fixture returns the correct routers without a project 1`] = `
[
  RedwoodRouter {
    "errors": [],
    "filepath": "/web/src/Routes.js",
    "name": "Routes",
    "routes": [
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "home",
        "pageIdentifier": "HomePage",
        "path": "/",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "typescriptPage",
        "pageIdentifier": "TypeScriptPage",
        "path": "/typescript",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "someOtherPage",
        "pageIdentifier": "EditUserPage",
        "path": "/somewhereElse",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "fooPage",
        "pageIdentifier": "FooPage",
        "path": "/foo",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "barPage",
        "pageIdentifier": "BarPage",
        "path": "/bar",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": true,
        "name": "privatePage",
        "pageIdentifier": "PrivatePage",
        "path": "/private-page",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": true,
        "isPrivate": false,
        "name": "notfound",
        "pageIdentifier": "NotFoundPage",
        "path": undefined,
        "prerender": false,
        "warnings": [],
      },
    ],
    "warnings": [],
  },
]
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct routers with a project 1`] = `
[
  RedwoodRouter {
    "errors": [
      "No notfound route detected",
      "Multiple routes named home are present",
    ],
    "filepath": "/web/src/Routes.js",
    "name": "Routes",
    "routes": [
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "home",
        "pageIdentifier": "HomePage",
        "path": "/",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "home",
        "pageIdentifier": "HomePage",
        "path": "/",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "something-else",
        "pageIdentifier": "DoesNotExist",
        "path": "/{foo}/{foo}",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "Routes",
        "pageIdentifier": "NotFoundPage",
        "path": undefined,
        "prerender": false,
        "warnings": [],
      },
    ],
    "warnings": [],
  },
]
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct routers without a project 1`] = `
[
  RedwoodRouter {
    "errors": [
      "No notfound route detected",
      "Multiple routes named home are present",
    ],
    "filepath": "/web/src/Routes.js",
    "name": "Routes",
    "routes": [
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "home",
        "pageIdentifier": "HomePage",
        "path": "/",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "home",
        "pageIdentifier": "HomePage",
        "path": "/",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "something-else",
        "pageIdentifier": "DoesNotExist",
        "path": "/{foo}/{foo}",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.js",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "Routes",
        "pageIdentifier": "NotFoundPage",
        "path": undefined,
        "prerender": false,
        "warnings": [],
      },
    ],
    "warnings": [],
  },
]
`;

exports[`From within the test-project fixture returns the correct routers with a project 1`] = `
[
  RedwoodRouter {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "name": "Routes",
    "routes": [
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "login",
        "pageIdentifier": "LoginPage",
        "path": "/login",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "signup",
        "pageIdentifier": "SignupPage",
        "path": "/signup",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "forgotPassword",
        "pageIdentifier": "ForgotPasswordPage",
        "path": "/forgot-password",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "resetPassword",
        "pageIdentifier": "ResetPasswordPage",
        "path": "/reset-password",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "newContact",
        "pageIdentifier": "ContactNewContactPage",
        "path": "/contacts/new",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "editContact",
        "pageIdentifier": "ContactEditContactPage",
        "path": "/contacts/{id:Int}/edit",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "contact",
        "pageIdentifier": "ContactContactPage",
        "path": "/contacts/{id:Int}",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "contacts",
        "pageIdentifier": "ContactContactsPage",
        "path": "/contacts",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "newPost",
        "pageIdentifier": "PostNewPostPage",
        "path": "/posts/new",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "editPost",
        "pageIdentifier": "PostEditPostPage",
        "path": "/posts/{id:Int}/edit",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "post",
        "pageIdentifier": "PostPostPage",
        "path": "/posts/{id:Int}",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "posts",
        "pageIdentifier": "PostPostsPage",
        "path": "/posts",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "waterfall",
        "pageIdentifier": "WaterfallPage",
        "path": "/waterfall/{id:Int}",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": true,
        "name": "profile",
        "pageIdentifier": "ProfilePage",
        "path": "/profile",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "blogPost",
        "pageIdentifier": "BlogPostPage",
        "path": "/blog-post/{id:Int}",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "contactUs",
        "pageIdentifier": "ContactUsPage",
        "path": "/contact",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "about",
        "pageIdentifier": "AboutPage",
        "path": "/about",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "home",
        "pageIdentifier": "HomePage",
        "path": "/",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": true,
        "isPrivate": false,
        "name": "notfound",
        "pageIdentifier": "NotFoundPage",
        "path": undefined,
        "prerender": true,
        "warnings": [],
      },
    ],
    "warnings": [],
  },
]
`;

exports[`From within the test-project fixture returns the correct routers without a project 1`] = `
[
  RedwoodRouter {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "name": "Routes",
    "routes": [
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "login",
        "pageIdentifier": "LoginPage",
        "path": "/login",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "signup",
        "pageIdentifier": "SignupPage",
        "path": "/signup",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "forgotPassword",
        "pageIdentifier": "ForgotPasswordPage",
        "path": "/forgot-password",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "resetPassword",
        "pageIdentifier": "ResetPasswordPage",
        "path": "/reset-password",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "newContact",
        "pageIdentifier": "ContactNewContactPage",
        "path": "/contacts/new",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "editContact",
        "pageIdentifier": "ContactEditContactPage",
        "path": "/contacts/{id:Int}/edit",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "contact",
        "pageIdentifier": "ContactContactPage",
        "path": "/contacts/{id:Int}",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "contacts",
        "pageIdentifier": "ContactContactsPage",
        "path": "/contacts",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "newPost",
        "pageIdentifier": "PostNewPostPage",
        "path": "/posts/new",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "editPost",
        "pageIdentifier": "PostEditPostPage",
        "path": "/posts/{id:Int}/edit",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "post",
        "pageIdentifier": "PostPostPage",
        "path": "/posts/{id:Int}",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "posts",
        "pageIdentifier": "PostPostsPage",
        "path": "/posts",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "waterfall",
        "pageIdentifier": "WaterfallPage",
        "path": "/waterfall/{id:Int}",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": true,
        "name": "profile",
        "pageIdentifier": "ProfilePage",
        "path": "/profile",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": true,
        "isNotFound": false,
        "isPrivate": false,
        "name": "blogPost",
        "pageIdentifier": "BlogPostPage",
        "path": "/blog-post/{id:Int}",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "contactUs",
        "pageIdentifier": "ContactUsPage",
        "path": "/contact",
        "prerender": false,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "about",
        "pageIdentifier": "AboutPage",
        "path": "/about",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": false,
        "isPrivate": false,
        "name": "home",
        "pageIdentifier": "HomePage",
        "path": "/",
        "prerender": true,
        "warnings": [],
      },
      RedwoodRoute {
        "errors": [],
        "filepath": "/web/src/Routes.tsx",
        "hasParameters": false,
        "isNotFound": true,
        "isPrivate": false,
        "name": "notfound",
        "pageIdentifier": "NotFoundPage",
        "path": undefined,
        "prerender": true,
        "warnings": [],
      },
    ],
    "warnings": [],
  },
]
`;
